sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/policy/ResponseValidationPolicyBuilder.java

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.storage.common.policy;
+
+import com.azure.core.http.HttpPipelineCallContext;
+import com.azure.core.http.HttpPipelineNextPolicy;
+import com.azure.core.http.HttpResponse;
+import com.azure.core.http.policy.HttpPipelinePolicy;
+import com.azure.core.util.logging.ClientLogger;
+import reactor.core.publisher.Mono;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.function.BiConsumer;
+
+/**
+ * Builder for a policy to do validation of general response behavior.
+ */
+public class ResponseValidationPolicyBuilder {
+
+    private final Collection<BiConsumer<HttpResponse, ClientLogger>> assertions = new ArrayList<>();
+
+    /**
+     * Builds the policy described by this builder.
+     *
+     * @return The policy.
+     */
+    public HttpPipelinePolicy build() {
+        return new ResponseValidationPolicy(assertions);
+    }
+
+    /**
+     * Fluently applies an optional validation to this policy where, if the response contains the given header, asserts
+     * its value is an echo of the value provided in the request.
+     *
+     * @param headerName The header to validate.
+     * @return This policy.
+     */
+    public ResponseValidationPolicyBuilder optionalEcho(String headerName) {
+        assertions.add((httpResponse, logger) -> {
+            String responseHeaderValue = httpResponse.headerValue(headerName);
+            if (responseHeaderValue != null && !responseHeaderValue.equals(httpResponse.request().headers().value(headerName))) {
```

Comment:
Does httpResponse.request().headers().value(headerName) throw if headerName is invalid?

Reply:
Javadocs say it returns null.

Reply:
Should be fine if it returns null